name-registry/src/main.sw

```diff
@@ -0,0 +1,154 @@
+            duration,
+            name,
+            new_expiry: new_record.expiry,
+        });
+    }
+
+    #[storage(read, write)]
+    fn register(
+        name: str[8],
+        duration: u64,
+        owner: Identity,
+        identity: Identity,
+    ) {
+        if storage.names.get(name).is_some() {
+            let record = storage.names.get(name).unwrap();
+            require(timestamp() > record.expiry, ValidityErrors::NameNotExpired);
+        }
+
+        require((duration / 100) * price_per_hundred <= msg_amount(), AssetErrors::InsufficientPayment);
+        require(msg_asset_id() == BASE_ASSET_ID, AssetErrors::IncorrectAssetSent);
+
+        let record = Record {
+            expiry: timestamp() + duration,
+            identity,
+            owner,
+        };
+
+        storage.names.insert(name, Option::Some(record));
+
+        log(NameRegisteredEvent {
+            expiry: record.expiry,
+            name,
+            owner,
+            identity,
+        });
+    }
+
+    #[storage(read, write)]
+    fn set_identity(name: str[8], identity: Identity) {
+        require(storage.names.get(name).is_some(), ValidityErrors::NameNotRegistered);
+        let previous_record = storage.names.get(name).unwrap();
+        require(timestamp() < previous_record.expiry, ValidityErrors::NameExpired);
+        require(previous_record.owner == msg_sender().unwrap(), AuthorisationErrors::SenderNotOwner);
+
+        let new_record = Record {
+            expiry: previous_record.expiry,
+            identity,
+            owner: previous_record.owner,
+        };
+
+        storage.names.insert(name, Option::Some(new_record));
+
+        log(IdentityChangedEvent {
+            name,
+            new_identity: new_record.identity,
+            previous_identity: previous_record.identity,
+        });
+    }
+
+    #[storage(read, write)]
+    fn set_owner(name: str[8], new_owner: Identity) {
+        require(storage.names.get(name).is_some(), ValidityErrors::NameNotRegistered);
+        let previous_record = storage.names.get(name).unwrap();
+        require(timestamp() < previous_record.expiry, ValidityErrors::NameExpired);
+        require(previous_record.owner == msg_sender().unwrap(), AuthorisationErrors::SenderNotOwner);
+
+        let new_record = Record {
+            expiry: previous_record.expiry,
+            identity: previous_record.identity,
+            owner: new_owner,
+        };
+
+        storage.names.insert(name, Option::Some(new_record));
+
+        log(OwnerChangedEvent {
+            name,
+            new_owner: new_record.owner,
+            previous_owner: previous_record.owner,
+        });
+    }
+
+    #[storage(read)]
+    fn expiry(name: str[8]) -> u64 {
+        require(storage.names.get(name).is_some(), ValidityErrors::NameNotRegistered);
+        let record = storage.names.get(name).unwrap();
+        require(timestamp() < record.expiry, ValidityErrors::NameExpired);
```

Comment:
I think the getter should not be reverting here when a record has expired. I think it should return 0. I think it should continue to revert in the two getters below though.

Alternatively, the 3 getters can return an option which is None if the record does not exist / is expired otherwise it returns whatever it needs to. I think that's probably a better approach for modularity / integration with contracts. This will allow them to handle logic rather than crash.
This also aligns with your spec which does not mention any revert if the name has expired.

Reply:
I dont think the SDK supports Options so testing would be impossible if I returned an Option